contracts/ring.md

```diff
@@ -0,0 +1,340 @@
+provides special special syntax for map/record updates.  Open
+question: we likely want to make it possible to have immutable state
+(parameters). Keep separate from mutable state or annotate certain
+fields as immutable?
+
+### Types
+Ring has the following types:
+
+| Type    | Description                     | Example
+| ------- | ------------------------------- | -------:
+| uint    | A 256 bit integer               | ```42```
+| int     | A 256 bit 2-complement integer  | ```-1```
+| address | A 256 bit number given as a hex | ```ff00```
+| bool    | A Boolean                       | ```true```
+| string  | An array of bytes               | ```“Foo”```
+| list    | A homogeneous immutable singly linked list. | ```[1, 2, 3]```
+| tuple   | An ordered heterogeneous array   | ```(42, “Foo”, true)```
+| record  | An immutable key value store with fixed key names and typed values | ``` type balance = { owner: address, value: uint } ```
+| map     | An immutable key value store with dynamic mapping of keys of one type to values of one type | ```type accounts = map(string, address)```
+| state   | A record of blockstate key, value pairs  |
+| transactions | An append only list of blockchain transactions |
+| events   | An append only list of blockchain events (or log entries) |
+
+#### Arrays
+Ring might not have arrays… Ring might have Maps.
+
+####  Variants
+Depending on how we write the compiler, if they are cheap to include
+some like to use variant types, while others might not want them.
+
+####  Refs
+ Ring has no refs.
+
+####  Object
+ Ring will not have objects. (only contracts) 
+
+### Pattern matching
+Pattern matching is probably outside the first iteration of Ring, but
+we definitely want it in the final language.
+
+### Builtins
+#### Events
+Ring has events which logs a structured message to the contract log in
+the resulting blockchain transaction.
+
+```
+event(name: string, value: any)
+```
+
+#### Transactions
+Ring can generate blockchain transactions.
+```
+transaction(tx_type: transaction, argument: record )
+```
+
+#### State
+
+A Ring contract has a state. You must declare the state record type in
+the contract. You can then update the state through mutable record
+update.
+
+```
+type state = { mutable field: string
+             , mutable value: uint};
+let state = {field: “”, value : 0};
+let store_value = (f: string, v: uint) => {
+    state.field = f;
+    state.value = v;
+}
+```
+#### Contract primitives
+```
+creator() : () => address
+```
+The address (pubkey) of the entity signing for the contract creation transaction
+```
+caller() : () => address
+```
+The address of the entity currently calling the contract.
+```
+value() : () => uint
+```
+The amount of coins transferred to the contract when calling.
+```
+balance() : () => uint
+```
+The amount of coins currently in the contract account
+```
+contract_address() : () => address;
+```
+The address of the contract.
+
+Block references
```

Comment:
Is it possible the access previous blocks?

Reply:
Yes, as far as the VM allows. I am not sure how to express that in the langugage.
What does Solidity do?

Reply:
Solidity provides a special, global variable `blocks`, which allows to get the hashes of the previous 256 blocks and also some properties of the current block, [see here](https://solidity.readthedocs.io/en/develop/units-and-global-variables.html#block-and-transaction-properties)